weave-js/src/components/PagePanelComponents/Home/Browse3/components/TraceNavigator/TraceViews/TreeView.tsx

```diff
@@ -0,0 +1,599 @@
+  | 'tool'
+  | 'llm'
+  | 'model'
+  | 'evaluation'
+  | 'scorer'
+  | 'none';
+
+// Helper function to get call type icon
+const getCallTypeIcon = (type: NodeType): IconName => {
+  switch (type) {
+    case 'agent':
+      return 'robot-service-member';
+    case 'tool':
+      return 'code-alt';
+    case 'llm':
+      return 'forum-chat-bubble';
+    case 'model':
+      return 'model';
+    case 'evaluation':
+      return 'baseline-alt';
+    case 'scorer':
+      return 'number';
+    case 'none':
+      return 'circle';
+    default:
+      return 'circle';
+  }
+};
+
+const opTypeToColor = (typeName: NodeType): string => {
+  switch (typeName) {
+    // Identifiers
+    case 'agent':
+    case 'model':
+      return 'text-blue-500 dark:text-blue-400';
+    // Evals
+    case 'tool':
+    case 'llm':
+      return 'text-magenta-600 dark:text-magenta-500';
+    // Evals
+    case 'evaluation':
+    case 'scorer':
+      return 'text-sienna-500 dark:text-sienna-400';
+    // Other, probable noise
+    default:
+      return 'text-moon-300 dark:text-moon-200';
+  }
+};
+
+// Note to future dev: this should probably be configurable at the database / attribute level
+// for now we use these simple heuristics but this can be improved
+const spanNameToTypeHeuristic = (spanName: string): NodeType => {
+  spanName = spanName.toLowerCase();
+  if (spanName.includes('agent')) {
+    return 'agent';
+  }
+  if (spanName.includes('tool')) {
+    return 'tool';
+  }
+  if (spanName.includes('score')) {
+    return 'scorer';
+  }
+  if (
+    spanName.includes('completion') ||
+    spanName.includes('generation') ||
+    spanName.includes('chat') ||
+    spanName.includes('llm')
+  ) {
+    return 'llm';
+  }
+  if (
+    spanName.includes('model') ||
+    spanName.includes('predict') ||
+    spanName.includes('generate') ||
+    spanName.includes('invoke')
+  ) {
+    return 'model';
+  }
+  if (spanName.includes('evaluation')) {
+    return 'evaluation';
+  }
+  return 'none';
+};
+
+const TreeNode: React.FC<TreeNodeProps> = ({
+  node,
+  style,
+  focusedCallId,
+  setFocusedCallId,
+  setRootCallId,
+  onToggleExpand,
+  deemphasizeCallIds,
+  searchQuery,
+  showDuration,
+}) => {
+  const {id, call, level, isExpanded, childrenIds, hasDescendantErrors} = node;
+  const duration = call.ended_at
+    ? Date.parse(call.ended_at) - Date.parse(call.started_at)
+    : null;
+
+  const spanName = parseSpanName(call.op_name);
+  const typeName = spanNameToTypeHeuristic(spanName);
+  const {cost, costToolTipContent} = getCostFromCostData(
+    call.summary?.weave?.costs
+  );
+  const {tokens, tokenToolTipContent} = getTokensFromUsage(call.summary?.usage);
+
+  const displayName = getCallDisplayName(call);
+  const renderHighlightedText = () => {
+    if (!searchQuery) {
+      return displayName;
+    }
+    const searchLower = searchQuery.toLowerCase();
+    const textLower = displayName.toLowerCase();
+    const index = textLower.indexOf(searchLower);
```

Comment:
I think ideally we'd highlight all matches within the name

Reply:
added note for followup here